### DISCUSSION TITLE
Inheritance and automatic downcasting does not work

### DISCUSSION THREAD

**Discussion by PreyMa** [wjakob/nanobind#125] [Q&A] [answered]
Hello, 
I noticed, that the recently added [documentation page](https://nanobind.readthedocs.io) mentions automatic downcasting capabilites for polymorphic types. It is described in detail [here](https://nanobind.readthedocs.io/en/latest/classes.html#inheritance-and-automatic-downcasting), even featuring an example.

However, I cannot get it to work. Even the provided example seems to be broken. I also tried the code from the [tests](https://github.com/wjakob/nanobind/blob/9d1f655b46e80c9d42c9943c5ecce0c0865f6c27/tests/test_classes.cpp) and added four simple functions, which also yield the parent class `Animal`. Function `test2` and `test3` should however return the child classes `Cat` and `Dog`, as far as I understand.

```c++
m.def("test0", []() { return std::unique_ptr<Cat>(new Cat("eine katze")); });
m.def("test1", []() { return std::unique_ptr<Dog>(new Dog("ein hund")); });
m.def("test2", []() { return std::unique_ptr<Animal>(new Cat("poly katze")); });
m.def("test3", []() { return std::unique_ptr<Animal>(new Dog("poly hund")); });
```

Could it be that this feature, which exists in [pybind11](https://github.com/pybind/pybind11), was cut from nanobind? [The documentation](https://pybind11.readthedocs.io/en/stable/classes.html?highlight=automatic%20down#inheritance-and-automatic-downcasting) looks pretty much identical, only with "pybind11" replaced by "nanobind" in the last paragraph. The link in the info box is also broken, which further indicates, that this might have been a copy-past mistake.

If I am not doing anything wrong and this feature is currently missing, I would be curious, if there was a plan to add it. Automatic downcasting was not mentioned on the list of removed pybind11 features. It seems pretty useful/important to me.
Thank you in advance.

**Comment by wjakob** [comment, marked as answer]
That's right. This part of the documentation was blatantly copied and still needs some massaging. You are right that this was one of the features that was cut in nanobind (and intentionally so, you should stay with pybind11 if you need it.)

**Reply by PreyMa** [reply]
Hello,
wow that was a quick reply, thank you! Although it would have been nice to have, I will work around this. Looks like I could have saved quite a bit of time if I asked earlier.
Again, thank you for the speedy response! :^)

**Reply by wjakob** [reply]
Hi @PreyMa  -- I ended up thinking a bit more about this and decided to reintroduce the feature in a way that doesn't add overheads for the general case. The feature is available on the `master` branch and will be part of the next release. See https://nanobind.readthedocs.io/en/latest/classes.html#automatic-downcasting

**Reply by PreyMa** [reply]
Hello @wjakob, I am happy this feature is back. I think it helps a lot in creating more friendly and explorable Python interfaces when using the REPL for example! Also, this addition should remove the need for manual down-casts and required type checks in Python. Limiting this to `nb::class<...>` seems like a fair limitation. :^)

**Reply by wjakob** [reply]
FWIW That was always a limitation, even in pybind11 ;). They should behave identically.
